File: xtask/mpmath_eval/functionals/tw.py
"""mpmath port of xcfun-master/src/functionals/tw.cpp at mp.prec=200.

Thomas-Weizsäcker kinetic-GGA functional. Plan 06-N2 Task 1c (kinetic-GGA
family). No upstream test_in (Phase 2 Plan 02-05); mpmath at prec=200 is
sole reference per D-03 ACC-04 amendment.

Formula (verbatim port of `tw.cpp:20-22`):
    tw(d) = (1/8) * (gaa + gbb)^2 / n
where n = a + b. Cross-checked against the Rust kernel at
`crates/xcfun-kernels/src/functionals/lda/tw.rs:39-50` (1:1 port —
ctaylor_add → ctaylor_powi_2 → ctaylor_reciprocal → ctaylor_mul →
ctaylor_scalar_mul(0.125)).
"""
from __future__ import annotations
import mpmath as mp

from ..ad_chain import multivariate_taylor
from ..densvars import slot_names


def _value_tw(*inputs, vars_str):
    slots = slot_names(vars_str)
    d = {name: mp.mpf(val) for name, val in zip(slots, inputs)}
    a = d.get("a", mp.mpf(0))
    b = d.get("b", mp.mpf(0))
    gaa = d.get("gaa", mp.mpf(0))
    gbb = d.get("gbb", mp.mpf(0))
    n = a + b
    return mp.mpf("0.125") * (gaa + gbb) ** 2 / n


def eval_tw(inputs, vars, mode, order):
    mp.mp.prec = 200
    pt = tuple(mp.mpf(x) for x in inputs)
    return multivariate_taylor(
        lambda *args, _vs=vars: _value_tw(*args, vars_str=_vs),
        pt,
        order,
    )

File: crates/xcfun-kernels/src/functionals/lda/tw.rs
//! Thomas-Weizsäcker kinetic energy functional. **LDA-09 part 2 (kinetic-GGA).**
//!
//! # Source
//! - `xcfun-master/src/functionals/tw.cpp:20-30` (depends `XC_DENSITY | XC_GRADIENT`)
//!
//! # Formula
//! $$ T_W = \frac{1}{8} \cdot (\text{gaa} + \text{gbb})^2 / n $$
//!
//! # Preconditions (Pitfall PHASE2-D)
//! - `d.gaa`, `d.gbb` populated by `build_densvars` `XC_A_B_GAA_GAB_GBB` arm (Plan 02-05 Wave-1C-1).
//! - The pure-density `XC_A_B` arm leaves `gaa = gbb = 0` — TW would silently
//!   return 0 in that case. Must be driven through the GAA_GAB_GBB builder.
//! - `d.n > 0` (regularize ensures `>= TINY_DENSITY = 1e-14`; division by zero guarded).

use cubecl::prelude::*;
use xcfun_ad::ctaylor::{ctaylor_add, ctaylor_scalar_mul};
use xcfun_ad::ctaylor_rec::mul::ctaylor_mul;
use xcfun_ad::math::{ctaylor_powi_2, ctaylor_reciprocal};

use crate::density_vars::DensVarsDev;

/// Thomas-Weizsäcker kinetic kernel. 1:1 port of `tw.cpp:20-22`:
///
/// ```cpp
/// return 1. / 8. * pow(d.gaa + d.gbb, 2.0) / d.n;
/// ```
///
/// Operation order (left-to-right, ACC-06 no mul_add):
///   1. sum = gaa + gbb                 (ctaylor_add)
///   2. sum2 = sum^2                    (ctaylor_powi_2 — fused x*x)
///   3. inv_n = 1/n                     (ctaylor_reciprocal)
///   4. tmp = sum2 * inv_n              (ctaylor_mul)
///   5. out = 0.125 * tmp               (ctaylor_scalar_mul; 1/8 = 0.125)
#[cube]
pub fn tw_kernel<F: Float>(d: &DensVarsDev<F>, out: &mut Array<F>, #[comptime] n: u32) {
    let mut sum = Array::<F>::new(comptime!((1_u32 << n) as usize));
    let mut sum2 = Array::<F>::new(comptime!((1_u32 << n) as usize));
    let mut inv_n = Array::<F>::new(comptime!((1_u32 << n) as usize));
    let mut tmp = Array::<F>::new(comptime!((1_u32 << n) as usize));

    ctaylor_add::<F>(&d.gaa, &d.gbb, &mut sum, n);
    ctaylor_powi_2::<F>(&sum, &mut sum2, n);
    ctaylor_reciprocal::<F>(&d.n, &mut inv_n, n);
    ctaylor_mul::<F>(&sum2, &inv_n, &mut tmp, n);
    ctaylor_scalar_mul::<F>(&tmp, F::cast_from(0.125_f64), out, n);
}
